File: Sample Applications/TelerikCheckers/TelerikCheckersCS/CheckersBoard.cs
using System;
using Telerik.WinControls.UI;
using System.Drawing;
using TelerikCheckers.Properties;
using Telerik.WinControls;

namespace TelerikCheckers
{
    public class CheckersBoard
    {
        public static Image BigBlackChecker = new Bitmap(Resources.checker_black, new Size(80, 80));
        public static Image BigWhiteChecker = new Bitmap(Resources.checker_white, new Size(80, 80));
        public static Image BigBlackCheckerKing = new Bitmap(Resources.checker_black_king, new Size(80, 80));
        public static Image BigWhiteCheckerKing = new Bitmap(Resources.checker_white_king, new Size(80, 80));
        public static Image BlackChecker = Resources.checker_black;
        public static Image WhiteChecker = Resources.checker_white;
        public static Image BlackCheckerKing = Resources.checker_black_king;
        public static Image WhiteCheckerKing = Resources.checker_white_king;

        public static Color BorderOuterColor = Color.FromArgb(58, 27, 8);
        public static Color BorderInnerColor = Color.FromArgb(248, 229, 196);
        public static Color FontColor = Color.FromArgb(248, 229, 196);
        public static Color BackDarkColor = Color.FromArgb(132, 62, 17);
        public static Color BackLightColor = Color.FromArgb(223, 177, 94);

        public static string WHITE = "W";
        public static string BLACK = "B";
        public static string WHITEKING = "WC";
        public static string BLACKKING = "BC";
        public static int whiteTaken = 0;
        public static int blackTaken = 0;
        public static int rowIndex = -1;
        public static int columnIndex = -1;
        public static bool shouldTakeChecker = false;

        public static bool CanCheckerBeDragged(Turn turn, GridDataCellElement cell)
        {
            string value = cell.Value.ToString();
            if ((value == WHITE && turn == Turn.White)
                || (value == BLACK && turn == Turn.Black)
                || (value == WHITEKING && turn == Turn.White)
                || (value == BLACKKING && turn == Turn.Black))
            {
                return true;
            }
            return false;
        }

        public static bool CanHitMoreCheckersInRow(Turn turn, GridDataCellElement targetCell)
        {
            RadGridView grid = targetCell.GridControl;

            if (turn == Turn.Black)
            {
                // top-right direction
                if (targetCell.RowIndex - 2 > -1 && targetCell.ColumnIndex + 2 < 8)
                {
                    object value = grid.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex + 1].Value;
                    if(value != null)
                    {
                        string stringValue = value.ToString();
                        if((stringValue == CheckersBoard.WHITE ||
                            stringValue == CheckersBoard.WHITEKING)
                            && grid.Rows[targetCell.RowIndex - 2].Cells[targetCell.ColumnIndex + 2].Value == null)
                        {
                            return true;
                        }
                    }
                }

                // top-left direction
                if (targetCell.RowIndex - 2 > -1 && targetCell.ColumnIndex - 2 > -1)
                {
                    object value = grid.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex - 1].Value;
                    if (value != null)
                    {
                        string stringValue = value.ToString();
                        if ((stringValue == CheckersBoard.WHITE ||
                            stringValue == CheckersBoard.WHITEKING)
                            && grid.Rows[targetCell.RowIndex - 2].Cells[targetCell.ColumnIndex - 2].Value == null)
                        {
                            return true;
                        }
                    }
                }

                if (targetCell.Value.ToString() == CheckersBoard.BLACKKING)
                {                  
                    // bottom-right direction
                    if (targetCell.RowIndex + 2 < 8 && targetCell.ColumnIndex + 2 < 8)
                    {
                        object value = grid.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (value != null)
                        {
                            string stringValue = value.ToString();
                            if ((stringValue == CheckersBoard.WHITE ||
                                stringValue == CheckersBoard.WHITEKING)
                                && grid.Rows[targetCell.RowIndex + 2].Cells[targetCell.ColumnIndex + 2].Value == null)
                            {
                                return true;
                            }
                        }
                    }

                    // bottom-left direction
                    if (targetCell.RowIndex + 2 < 8 && targetCell.ColumnIndex - 2 > -1)
                    {

                        object value = grid.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (value != null)
                        {
                            string stringValue = value.ToString();
                            if ((stringValue == CheckersBoard.WHITE ||
                                stringValue == CheckersBoard.WHITEKING)
                                && grid.Rows[targetCell.RowIndex + 2].Cells[targetCell.ColumnIndex - 1].Value == null)
                            {
                                return true;
                            }
                        }
                    }
                }
            }
            else
            {
                // bottom-right direction
                if (targetCell.RowIndex + 2 < 8 && targetCell.ColumnIndex + 2 < 8)
                {

                    object value = grid.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex + 1].Value;
                    if (value != null)
                    {
                        string stringValue = value.ToString();
                        if ((stringValue == CheckersBoard.WHITE ||
                            stringValue == CheckersBoard.WHITEKING)
                            && grid.Rows[targetCell.RowIndex + 2].Cells[targetCell.ColumnIndex + 2].Value == null)
                        {
                            return true;
                        }
                    }
                }

                // bottom-left direction
                if (targetCell.RowIndex + 2 < 8 && targetCell.ColumnIndex - 2 > -1)
                {
                    object value = grid.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex - 1].Value;
                    if (value != null)
                    {
                        string stringValue = value.ToString();
                        if ((stringValue == CheckersBoard.WHITE ||
                            stringValue == CheckersBoard.WHITEKING)
                            && grid.Rows[targetCell.RowIndex + 2].Cells[targetCell.ColumnIndex - 1].Value == null)
                        {
                            return true;
                        }
                    }
                }

                if (targetCell.Value.ToString() == CheckersBoard.WHITEKING)
                {
                    // top-right direction
                    if (targetCell.RowIndex - 2 > -1 && targetCell.ColumnIndex + 2 < 8)
                    {
                        object value = grid.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (value != null)
                        {
                            string stringValue = value.ToString();
                            if ((stringValue == CheckersBoard.WHITE ||
                                stringValue == CheckersBoard.WHITEKING)
                                && grid.Rows[targetCell.RowIndex - 2].Cells[targetCell.ColumnIndex + 2].Value == null)
                            {
                                return true;
                            }
                        }
                    }

                    // top-left direction
                    if (targetCell.RowIndex - 2 > -1 && targetCell.ColumnIndex - 2 > -1)
                    {
                        object value = grid.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (value != null)
                        {
                            string stringValue = value.ToString();
                            if ((stringValue == CheckersBoard.WHITE ||
                                stringValue == CheckersBoard.WHITEKING)
                                && grid.Rows[targetCell.RowIndex - 2].Cells[targetCell.ColumnIndex - 1].Value == null)
                            {
                                return true;
                            }
                        }
                    }
                }
            }

            return false;
        }

        public static bool IsLastRow(GridDataCellElement targetCell)
        {
            if (targetCell.RowIndex == 0 || targetCell.RowIndex == 7)
            {
                return true;
            }

            return false;
        }

        public static bool IsCheckerWhite(string checker)
        {
            if (checker == WHITE
                || checker == WHITEKING)
            {
                return true;
            }
            return false;
        }

        public static bool IsCheckerBlack(string checker)
        {
            if (checker == BLACK
                || checker == BLACKKING)
            {
                return true;
            }
            return false;
        }

        public static bool CanDrop(GridDataCellElement targetCell, GridDataCellElement draggedCell, string draggedValue)
        {
            rowIndex = -1;
            columnIndex = -1;

            // you can't drag one checker on another
            if (targetCell.Value != null)
            {
                return false;
            }

            // you can't drag a checker on a white square
            if ((targetCell.RowIndex + targetCell.ColumnIndex) % 2 == 0)
            {
                return false;
            }

            if (!CheckersBoard.shouldTakeChecker)
            {
                // move only up
                if (draggedValue == BLACK)
                {
                    if (draggedCell.RowIndex - targetCell.RowIndex == 1
                        && Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) == 1)
                    {
                        return true;
                    }
                }

                // move only down
                if (draggedValue == WHITE)
                {
                    if (targetCell.RowIndex - draggedCell.RowIndex == 1
                        && Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) == 1)
                    {
                        return true;
                    }
                }

                // king can be moved up and down
                if (draggedValue == BLACKKING)
                {
                    if (Math.Abs(draggedCell.RowIndex - targetCell.RowIndex) == 1
                        && Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) == 1)
                    {
                        return true;
                    }
                }

                // king can be moved up and down
                if (draggedValue == WHITEKING)
                {
                    if (Math.Abs(draggedCell.RowIndex - targetCell.RowIndex) == 1
                        && Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) == 1)
                    {
                        return true;
                    }
                }
            }

            // if black checker is dragged over a white (or white king) checker
            if (draggedValue == BLACK)
            {
                // top
                if (targetCell.RowIndex - draggedCell.RowIndex == -2)
                {
                    // top-right possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == 2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue == WHITE || stringValue == WHITEKING)
                            {
                                rowIndex = targetCell.RowIndex + 1;
                                columnIndex = targetCell.ColumnIndex - 1;
                                return true;
                            }
                        }
                    }

                    // top-left possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == -2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue != null)
                            {
                                if (stringValue == WHITE || stringValue == WHITEKING)
                                {
                                    rowIndex = targetCell.RowIndex + 1;
                                    columnIndex = targetCell.ColumnIndex + 1;
                                    return true;
                                }
                            }
                        }
                    }
                }
            }

            // if white checker is dragged over a black (or black king) checker
            if (draggedValue == WHITE)
            {
                // top
                if (targetCell.RowIndex - draggedCell.RowIndex == 2)
                {
                    // bottom-right possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == 2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue == BLACK || stringValue == BLACKKING)
                            {
                                rowIndex = targetCell.RowIndex - 1;
                                columnIndex = targetCell.ColumnIndex - 1;
                                return true;
                            }
                        }
                    }

                    // bottom-left possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == -2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue == BLACK || stringValue == BLACKKING)
                            {
                                rowIndex = targetCell.RowIndex - 1;
                                columnIndex = targetCell.ColumnIndex + 1;
                                return true;
                            }
                        }
                    }
                }
            }

            // if black king checker is dragged over a white (or white king) checker
            if (draggedValue == BLACKKING)
            {
                // to-bottom
                if (targetCell.RowIndex - draggedCell.RowIndex == 2)
                {
                    // bottom-right possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == 2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue == WHITE || stringValue == WHITEKING)
                            {
                                rowIndex = targetCell.RowIndex - 1;
                                columnIndex = targetCell.ColumnIndex - 1;
                                return true;
                            }
                        }
                    }

                    // bottom-left possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == -2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (val != null)
                        {
                            string objectValue = val.ToString();
                            if (objectValue == WHITE || objectValue == WHITEKING)
                            {
                                rowIndex = targetCell.RowIndex - 1;
                                columnIndex = targetCell.ColumnIndex + 1;
                                return true;
                            }
                        }
                    }
                }

                // to-top
                if (targetCell.RowIndex - draggedCell.RowIndex == -2)
                {
                    // top-right possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == 2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue == WHITE || stringValue == WHITEKING)
                            {
                                rowIndex = targetCell.RowIndex + 1;
                                columnIndex = targetCell.ColumnIndex - 1;
                                return true;
                            }
                        }
                    }

                    // top-left possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == -2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (val != null)
                        {
                            string objectValue = val.ToString();
                            if (objectValue == WHITE || objectValue == WHITEKING)
                            {
                                rowIndex = targetCell.RowIndex + 1;
                                columnIndex = targetCell.ColumnIndex + 1;
                                return true;
                            }
                        }
                    }
                }
            }

            // if white checker is dragged over a black (or black king) checker
            if (draggedValue == WHITEKING)
            {
                // to-bottom
                if (targetCell.RowIndex - draggedCell.RowIndex == 2)
                {
                    // bottom-right possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == 2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue == BLACK || stringValue == BLACKKING)
                            {
                                rowIndex = targetCell.RowIndex - 1;
                                columnIndex = targetCell.ColumnIndex - 1;
                                return true;
                            }
                        }
                    }

                    // bottom-left possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == -2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex - 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (val != null)
                        {
                            string objectValue = val.ToString();
                            if (objectValue == BLACK || objectValue == BLACKKING)
                            {
                                rowIndex = targetCell.RowIndex - 1;
                                columnIndex = targetCell.ColumnIndex + 1;
                                return true;
                            }
                        }
                    }
                }

                // to-top
                if (targetCell.RowIndex - draggedCell.RowIndex == -2)
                {
                    // top-right possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == 2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex - 1].Value;
                        if (val != null)
                        {
                            string stringValue = val.ToString();
                            if (stringValue == BLACK || stringValue == BLACKKING)
                            {
                                rowIndex = targetCell.RowIndex + 1;
                                columnIndex = targetCell.ColumnIndex - 1;
                                return true;
                            }
                        }
                    }

                    // top-left possible square
                    if (targetCell.ColumnIndex - draggedCell.ColumnIndex == -2)
                    {
                        object val = targetCell.ViewInfo.Rows[targetCell.RowIndex + 1].Cells[targetCell.ColumnIndex + 1].Value;
                        if (val != null)
                        {
                            string objectValue = val.ToString();
                            if (objectValue == BLACK || objectValue == WHITEKING)
                            {
                                rowIndex = targetCell.RowIndex + 1;
                                columnIndex = targetCell.ColumnIndex + 1;
                                return true;
                            }
                        }
                    }
                }
            }
            return false;
        }

        public static void IsWinner(Turn turn)
        {
            if (turn == Turn.Black)
            {
                if (blackTaken == 12)
                {
                    RadMessageBox.Show("White wins!");
                }
            }
            else
            {
                if (whiteTaken == 12)
                {
                    RadMessageBox.Show("Black wins!");
                }
            }
        }
    }
}

File: Sample Applications/TelerikCheckers/TelerikCheckersVB/CheckersBoard.vb
Imports Microsoft.VisualBasic
Imports System
Imports Telerik.WinControls.UI
Imports System.Drawing
Imports TelerikCheckers.Properties
Imports Telerik.WinControls

Namespace TelerikCheckers
	Public Class CheckersBoard
		Public Shared BigBlackChecker As Image = New Bitmap(Resources.checker_black, New Size(80, 80))
		Public Shared BigWhiteChecker As Image = New Bitmap(Resources.checker_white, New Size(80, 80))
		Public Shared BigBlackCheckerKing As Image = New Bitmap(Resources.checker_black_king, New Size(80, 80))
		Public Shared BigWhiteCheckerKing As Image = New Bitmap(Resources.checker_white_king, New Size(80, 80))
		Public Shared BlackChecker As Image = Resources.checker_black
		Public Shared WhiteChecker As Image = Resources.checker_white
		Public Shared BlackCheckerKing As Image = Resources.checker_black_king
		Public Shared WhiteCheckerKing As Image = Resources.checker_white_king

		Public Shared BorderOuterColor As Color = Color.FromArgb(58, 27, 8)
		Public Shared BorderInnerColor As Color = Color.FromArgb(248, 229, 196)
		Public Shared FontColor As Color = Color.FromArgb(248, 229, 196)
		Public Shared BackDarkColor As Color = Color.FromArgb(132, 62, 17)
		Public Shared BackLightColor As Color = Color.FromArgb(223, 177, 94)

		Public Shared WHITE As String = "W"
		Public Shared BLACK As String = "B"
		Public Shared WHITEKING As String = "WC"
		Public Shared BLACKKING As String = "BC"
		Public Shared whiteTaken As Integer = 0
		Public Shared blackTaken As Integer = 0
		Public Shared rowIndex As Integer = -1
		Public Shared columnIndex As Integer = -1
		Public Shared shouldTakeChecker As Boolean = False

		Public Shared Function CanCheckerBeDragged(ByVal turn As Turn, ByVal cell As GridDataCellElement) As Boolean
			Dim value As String = cell.Value.ToString()
			If (value = WHITE AndAlso turn = Turn.White) OrElse (value = BLACK AndAlso turn = Turn.Black) OrElse (value = WHITEKING AndAlso turn = Turn.White) OrElse (value = BLACKKING AndAlso turn = Turn.Black) Then
				Return True
			End If
			Return False
		End Function

		Public Shared Function CanHitMoreCheckersInRow(ByVal turn As Turn, ByVal targetCell As GridDataCellElement) As Boolean
			Dim grid As RadGridView = targetCell.GridControl

			If turn = Turn.Black Then
				' top-right direction
				If targetCell.RowIndex - 2 > -1 AndAlso targetCell.ColumnIndex + 2 < 8 Then
					Dim value As Object = grid.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex + 1).Value
					If Not value Is Nothing Then
						Dim stringValue As String = value.ToString()
						If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex - 2).Cells(targetCell.ColumnIndex + 2).Value Is Nothing Then
							Return True
						End If
					End If
				End If

				' top-left direction
				If targetCell.RowIndex - 2 > -1 AndAlso targetCell.ColumnIndex - 2 > -1 Then
					Dim value As Object = grid.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex - 1).Value
					If Not value Is Nothing Then
						Dim stringValue As String = value.ToString()
						If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex - 2).Cells(targetCell.ColumnIndex - 2).Value Is Nothing Then
							Return True
						End If
					End If
				End If

				If targetCell.Value.ToString() = CheckersBoard.BLACKKING Then
					' bottom-right direction
					If targetCell.RowIndex + 2 < 8 AndAlso targetCell.ColumnIndex + 2 < 8 Then
						Dim value As Object = grid.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not value Is Nothing Then
							Dim stringValue As String = value.ToString()
							If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex + 2).Cells(targetCell.ColumnIndex + 2).Value Is Nothing Then
								Return True
							End If
						End If
					End If

					' bottom-left direction
					If targetCell.RowIndex + 2 < 8 AndAlso targetCell.ColumnIndex - 2 > -1 Then

						Dim value As Object = grid.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not value Is Nothing Then
							Dim stringValue As String = value.ToString()
							If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex + 2).Cells(targetCell.ColumnIndex - 1).Value Is Nothing Then
								Return True
							End If
						End If
					End If
				End If
			Else
				' bottom-right direction
				If targetCell.RowIndex + 2 < 8 AndAlso targetCell.ColumnIndex + 2 < 8 Then

					Dim value As Object = grid.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex + 1).Value
					If Not value Is Nothing Then
						Dim stringValue As String = value.ToString()
						If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex + 2).Cells(targetCell.ColumnIndex + 2).Value Is Nothing Then
							Return True
						End If
					End If
				End If

				' bottom-left direction
				If targetCell.RowIndex + 2 < 8 AndAlso targetCell.ColumnIndex - 2 > -1 Then
					Dim value As Object = grid.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex - 1).Value
					If Not value Is Nothing Then
						Dim stringValue As String = value.ToString()
						If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex + 2).Cells(targetCell.ColumnIndex - 1).Value Is Nothing Then
							Return True
						End If
					End If
				End If

				If targetCell.Value.ToString() = CheckersBoard.WHITEKING Then
					' top-right direction
					If targetCell.RowIndex - 2 > -1 AndAlso targetCell.ColumnIndex + 2 < 8 Then
						Dim value As Object = grid.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not value Is Nothing Then
							Dim stringValue As String = value.ToString()
							If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex - 2).Cells(targetCell.ColumnIndex + 2).Value Is Nothing Then
								Return True
							End If
						End If
					End If

					' top-left direction
					If targetCell.RowIndex - 2 > -1 AndAlso targetCell.ColumnIndex - 2 > -1 Then
						Dim value As Object = grid.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not value Is Nothing Then
							Dim stringValue As String = value.ToString()
							If (stringValue = CheckersBoard.WHITE OrElse stringValue = CheckersBoard.WHITEKING) AndAlso grid.Rows(targetCell.RowIndex - 2).Cells(targetCell.ColumnIndex - 1).Value Is Nothing Then
								Return True
							End If
						End If
					End If
				End If
			End If

			Return False
		End Function

		Public Shared Function IsLastRow(ByVal targetCell As GridDataCellElement) As Boolean
			If targetCell.RowIndex = 0 OrElse targetCell.RowIndex = 7 Then
				Return True
			End If

			Return False
		End Function

		Public Shared Function IsCheckerWhite(ByVal checker As String) As Boolean
			If checker = WHITE OrElse checker = WHITEKING Then
				Return True
			End If
			Return False
		End Function

		Public Shared Function IsCheckerBlack(ByVal checker As String) As Boolean
			If checker = BLACK OrElse checker = BLACKKING Then
				Return True
			End If
			Return False
		End Function

		Public Shared Function CanDrop(ByVal targetCell As GridDataCellElement, ByVal draggedCell As GridDataCellElement, ByVal draggedValue As String) As Boolean
			rowIndex = -1
			columnIndex = -1

			' you can't drag one checker on another
			If Not targetCell.Value Is Nothing Then
				Return False
			End If

			' you can't drag a checker on a white square
			If (targetCell.RowIndex + targetCell.ColumnIndex) Mod 2 = 0 Then
				Return False
			End If

			If (Not CheckersBoard.shouldTakeChecker) Then
				' move only up
				If draggedValue = BLACK Then
					If draggedCell.RowIndex - targetCell.RowIndex = 1 AndAlso Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) = 1 Then
						Return True
					End If
				End If

				' move only down
				If draggedValue = WHITE Then
					If targetCell.RowIndex - draggedCell.RowIndex = 1 AndAlso Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) = 1 Then
						Return True
					End If
				End If

				' king can be moved up and down
				If draggedValue = BLACKKING Then
					If Math.Abs(draggedCell.RowIndex - targetCell.RowIndex) = 1 AndAlso Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) = 1 Then
						Return True
					End If
				End If

				' king can be moved up and down
				If draggedValue = WHITEKING Then
					If Math.Abs(draggedCell.RowIndex - targetCell.RowIndex) = 1 AndAlso Math.Abs(draggedCell.ColumnIndex - targetCell.ColumnIndex) = 1 Then
						Return True
					End If
				End If
			End If

			' if black checker is dragged over a white (or white king) checker
			If draggedValue = BLACK Then
				' top
				If targetCell.RowIndex - draggedCell.RowIndex = -2 Then
					' top-right possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = 2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If stringValue = WHITE OrElse stringValue = WHITEKING Then
								rowIndex = targetCell.RowIndex + 1
								columnIndex = targetCell.ColumnIndex - 1
								Return True
							End If
						End If
					End If

					' top-left possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = -2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If Not stringValue Is Nothing Then
								If stringValue = WHITE OrElse stringValue = WHITEKING Then
									rowIndex = targetCell.RowIndex + 1
									columnIndex = targetCell.ColumnIndex + 1
									Return True
								End If
							End If
						End If
					End If
				End If
			End If

			' if white checker is dragged over a black (or black king) checker
			If draggedValue = WHITE Then
				' top
				If targetCell.RowIndex - draggedCell.RowIndex = 2 Then
					' bottom-right possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = 2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If stringValue = BLACK OrElse stringValue = BLACKKING Then
								rowIndex = targetCell.RowIndex - 1
								columnIndex = targetCell.ColumnIndex - 1
								Return True
							End If
						End If
					End If

					' bottom-left possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = -2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If stringValue = BLACK OrElse stringValue = BLACKKING Then
								rowIndex = targetCell.RowIndex - 1
								columnIndex = targetCell.ColumnIndex + 1
								Return True
							End If
						End If
					End If
				End If
			End If

			' if black king checker is dragged over a white (or white king) checker
			If draggedValue = BLACKKING Then
				' to-bottom
				If targetCell.RowIndex - draggedCell.RowIndex = 2 Then
					' bottom-right possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = 2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If stringValue = WHITE OrElse stringValue = WHITEKING Then
								rowIndex = targetCell.RowIndex - 1
								columnIndex = targetCell.ColumnIndex - 1
								Return True
							End If
						End If
					End If

					' bottom-left possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = -2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not val Is Nothing Then
							Dim objectValue As String = val.ToString()
							If objectValue = WHITE OrElse objectValue = WHITEKING Then
								rowIndex = targetCell.RowIndex - 1
								columnIndex = targetCell.ColumnIndex + 1
								Return True
							End If
						End If
					End If
				End If

				' to-top
				If targetCell.RowIndex - draggedCell.RowIndex = -2 Then
					' top-right possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = 2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If stringValue = WHITE OrElse stringValue = WHITEKING Then
								rowIndex = targetCell.RowIndex + 1
								columnIndex = targetCell.ColumnIndex - 1
								Return True
							End If
						End If
					End If

					' top-left possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = -2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not val Is Nothing Then
							Dim objectValue As String = val.ToString()
							If objectValue = WHITE OrElse objectValue = WHITEKING Then
								rowIndex = targetCell.RowIndex + 1
								columnIndex = targetCell.ColumnIndex + 1
								Return True
							End If
						End If
					End If
				End If
			End If

			' if white checker is dragged over a black (or black king) checker
			If draggedValue = WHITEKING Then
				' to-bottom
				If targetCell.RowIndex - draggedCell.RowIndex = 2 Then
					' bottom-right possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = 2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If stringValue = BLACK OrElse stringValue = BLACKKING Then
								rowIndex = targetCell.RowIndex - 1
								columnIndex = targetCell.ColumnIndex - 1
								Return True
							End If
						End If
					End If

					' bottom-left possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = -2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex - 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not val Is Nothing Then
							Dim objectValue As String = val.ToString()
							If objectValue = BLACK OrElse objectValue = BLACKKING Then
								rowIndex = targetCell.RowIndex - 1
								columnIndex = targetCell.ColumnIndex + 1
								Return True
							End If
						End If
					End If
				End If

				' to-top
				If targetCell.RowIndex - draggedCell.RowIndex = -2 Then
					' top-right possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = 2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex - 1).Value
						If Not val Is Nothing Then
							Dim stringValue As String = val.ToString()
							If stringValue = BLACK OrElse stringValue = BLACKKING Then
								rowIndex = targetCell.RowIndex + 1
								columnIndex = targetCell.ColumnIndex - 1
								Return True
							End If
						End If
					End If

					' top-left possible square
					If targetCell.ColumnIndex - draggedCell.ColumnIndex = -2 Then
						Dim val As Object = targetCell.ViewInfo.Rows(targetCell.RowIndex + 1).Cells(targetCell.ColumnIndex + 1).Value
						If Not val Is Nothing Then
							Dim objectValue As String = val.ToString()
							If objectValue = BLACK OrElse objectValue = WHITEKING Then
								rowIndex = targetCell.RowIndex + 1
								columnIndex = targetCell.ColumnIndex + 1
								Return True
							End If
						End If
					End If
				End If
			End If
			Return False
		End Function

		Public Shared Sub IsWinner(ByVal turn As Turn)
			If turn = Turn.Black Then
				If blackTaken = 12 Then
					RadMessageBox.Show("White wins!")
				End If
			Else
				If whiteTaken = 12 Then
					RadMessageBox.Show("Black wins!")
				End If
			End If
		End Sub
	End Class
End Namespace
